lib/ringbuf/src/lib.rs

```diff
@@ -282,33 +282,50 @@ pub struct Ringbuf<T: Copy + PartialEq, const N: usize> {
 
 impl<T: Copy + PartialEq, const N: usize> Ringbuf<T, { N }> {
     pub fn entry(&mut self, line: u16, payload: T) {
-        let ndx = match self.last {
-            None => 0,
-            Some(last) => {
-                let ent = &mut self.buffer[last];
+        // If this is the first time this ringbuf has been poked, last will be
+        // None. In this specific case we want to make sure we don't add to the
+        // count of an existing entry, and also that we deposit the first entry
+        // in slot 0. From a code generation perspective, the cheapest thing to
+        // do is to treat None as an out-of-range value:
+        let last = self.last.unwrap_or(usize::MAX);
 
-                if ent.line == line && ent.payload == payload {
-                    // Only reuse this entry if we don't overflow the
-                    // count.
-                    if let Some(new_count) = ent.count.checked_add(1) {
-                        ent.count = new_count;
-                        return;
-                    }
+        // Check to see if we can reuse the most recent entry. This uses get_mut
+        // both to avoid checking an entry on the first insertion (see above),
+        // and also to handle the case where last is somehow corrupted to point
+        // out-of-range. This avoids a bounds check panic. In the event that
+        // last _is_ corrupted, the behavior below will just start us over at 0.
+        if let Some(ent) = self.buffer.get_mut(last) {
+            if ent.line == line && ent.payload == payload {
+                // Only reuse this entry if we don't overflow the
+                // count.
+                if let Some(new_count) = ent.count.checked_add(1) {
+                    ent.count = new_count;
+                    return;
                 }
+            }
+        }
 
-                if last + 1 >= self.buffer.len() {
-                    0
-                } else {
-                    last + 1
-                }
+        // Either we were unable to reuse the entry, or the last index was out
+        // of range (perhaps because this is the first insertion). We're going
+        // to advance last and wrap if required. This uses a wrapping_add
+        // because if last is usize::MAX already, we want it to wrap to zero
+        // regardless -- and this avoids a checked arithmetic panic on the +1.
+        let ndx = {
+            let last_plus_1 = last.wrapping_add(1);
+            if last_plus_1 >= self.buffer.len() {
```

Comment:
Is the generated code quality still very poor when using modulus here?  This would be cleaner if it were just,

`let ndx = last.wrapping_add(1) % self.buffer.len();`

But if the generated code for that is still excessively verbose, it may be worth adding a comment about why this is written the way that it is.

Reply:
Out of curiousity, i tested this in Compiler Explorer, and unfortunately, both the `%` operator and `.wrapping_rem(self.buffer.len())` seem to generate _substantially_ more code than the `if` (although `wrapping_rem` is a little bit shorter). Since both `%` and `wrapping_rem` panic if the divisor is 0, they both generate a bunch of panic formatting code. 

See [this godbolt](https://godbolt.org/z/ex6Mv7G6s) (which might be nice to reference in the comment, actually!) for details.

Reply:
The genesis of the size concern is based on what this compiles to on M0, which lacks hardware division; the result there (at least as of a few months or a year ago or so, according to a Matrix post from Cliff: https://matrix.to/#/!jxnlDLnJsVaIRstnrt:oxide.computer/$li8Y-dTHnOVSdzi9z1n0tBqU-4xq9OXGLCahueefYXw?via=oxide.computer&via=unix.house&via=matrix.org) is 199 (!!) instructions. By comparison, any of these is significantly better.

The example for `wrapping_rem` on godbolt seemed incorrect (it's written as `wrapping_rem(1) % buflen`), but when I changed it to, `last.wrapping_add(1).wrapping_rem(buflen)` it appeared to generate the exact same code as the bare `%` version.

FWIW, the shortest version (at least for M$n$ where $n > 0$) seems to be, `unsafe { core::intrinsics::unchecked_rem(last.wrapping_add(1), buflen) }`, but that is terrible for other reasons.

In this case, it would be interesting to see what this compiles to in-situ; it wouldn't surprise me if the compiler couldn't figure out that `self.buffer.len()` is always > 0. My testing suggests that, if the compiler can figure out that the divisor is positive, it elides the panic check and associated support code.

Reply:
Ah, yeah, I see that if I change the target triple in the godbolt to the correct one for Cortex-M0 (`thumbv6m-none-eabi`), both the `%` and `wrapping_rem` (which I fixed) generate a call to `__aeabi_uidivmod`, which I assume is where the 199 instructions live (the Matrix link doesn't work for me; I'm guessing because I wasn't in that channel at the time). Here's my revised godbolt: https://godbolt.org/z/sac9PhzWx if it's useful.

Reply:
Can add a comment -- the reason I didn't use modulus here is two-fold after this change. First, we don't have hardware modulus (and on the M0, as you note, we don't even have hard divide). Second, it would require my `usize::MAX`-becomes-0 hack to be smarter, since `usize::MAX % n` is generally not zero.

Reply:
Not to flog a dead horse, but I was kind of curious just how bad this still was, so on a lark I played around with it for a few minutes. My testing was against app/nucleo-stm32g0/app-g070.toml and I compared generated code from el `jefe`.

One thing I noticed that this function is generic over `const N: usize`, and that `N` is the length of `self.buffer`, but is also a compile-time constant. I was curious what code would be generated for `let ndx = last.wrapping_add(1) % N;` --- presumably the optimizer can do a somewhat better job, since it knows that N's exact value, and that N>0. However, the compiler is too smart for that and both this case and the obvious case of `let ndx = last.wrapping_add(1) % self.buffer.len();` generated identical code.  One presumes that the (constant) size of the ring is propagated through compilation.

Interestingly, when ringbuf sizes are powers of two (as they are pretty much everywhere), the compiler does a really good job of optimizing things, and indeed, it generated largely the same code for both modulus versions and the explicit `if` construction (for some reason, r4 and r5 are swapped between the two versions and they are of course different, but interestingly seem to be the same size).  Here's the most interesting part of the diff:

```diff
-     162: 1c6c         	adds	r4, r5, #0x1
-     164: 2c03         	cmp	r4, #0x3
-     166: d900         	bls	0x16a <ringbuf::Ringbuf$LT$T$C$_$GT$::entry::habadc9fb7659bc36+0x8e> @ imm = #0x0
-     168: 2400         	movs	r4, #0x0
-     16a: 4363         	muls	r3, r4, r3
-     16c: 18ca         	adds	r2, r1, r3
-     16e: 9b02         	ldr	r3, [sp, #0x8]
-     170: 6113         	str	r3, [r2, #0x10]
-     172: 8190         	strh	r0, [r2, #0xc]
-     174: 89d0         	ldrh	r0, [r2, #0xe]
-     176: 1c40         	adds	r0, r0, #0x1
-     178: 81d0         	strh	r0, [r2, #0xe]
-     17a: 2001         	movs	r0, #0x1
-     17c: 6090         	str	r0, [r2, #0x8]
-     17e: c111         	stm	r1!, {r0, r4}
+     162: 2201         	movs	r2, #0x1
+     164: 1c64         	adds	r4, r4, #0x1
+     166: 2503         	movs	r5, #0x3
+     168: 4025         	ands	r5, r4
+     16a: c124         	stm	r1!, {r2, r5}
+     16c: 436b         	muls	r3, r5, r3
+     16e: 3908         	subs	r1, #0x8
+     170: 18c9         	adds	r1, r1, r3
+     172: 9b02         	ldr	r3, [sp, #0x8]
+     174: 610b         	str	r3, [r1, #0x10]
+     176: 8188         	strh	r0, [r1, #0xc]
+     178: 608a         	str	r2, [r1, #0x8]
+     17a: 89c8         	ldrh	r0, [r1, #0xe]
+     17c: 1c40         	adds	r0, r0, #0x1
+     17e: 81c8         	strh	r0, [r1, #0xe]
```

On the other hand, if the buffer size is _not_ a power of two, the compiler does a very poor job at code generation, decaying to the div/rem intrinsics in software. Since nothing says that the ring buffer sizes need be powers of two, that could easily happen.

Anyway, it was an interesting digression for an hour.